apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/signature/SignatureParser.java

```diff
@@ -0,0 +1,132 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.jdbc.signature;
+
+import static co.elastic.apm.agent.jdbc.signature.Scanner.Token.EOF;
+import static co.elastic.apm.agent.jdbc.signature.Scanner.Token.FROM;
+import static co.elastic.apm.agent.jdbc.signature.Scanner.Token.IDENT;
+import static co.elastic.apm.agent.jdbc.signature.Scanner.Token.LPAREN;
+import static co.elastic.apm.agent.jdbc.signature.Scanner.Token.RPAREN;
+
+public class SignatureParser {
+
+    private final Scanner scanner = new Scanner();
+
+    public void querySignature(String query, StringBuilder signature) {
```

Comment:
Shouldn't we cache the resulted name for a given query string? It is likely that the same string object will be reused, but even if not, I assume hash code calculation is faster than this scan (can be benchmarked).

Reply:
Good idea. But we also need to limit the size of the cache. When using algorithms like LRU for the cache and if there's a lot of dynamically generated SQL there will be a lot of thrashing and contention on that cache. I suspect that the contention is more expensive than just scanning the SQL. So what do you think about just caching the first ~200 SQL statements?

/cc @axw

Reply:
Some challenges: We should probably use a map with weak keys so that we don't leak strings which would otherwise not be referenced as it takes away space in the cache and the same query may not be used again afterwards. However, if we do that for an application where a large portion of SQLs are dynamically created, we end up allocating a lot of strings for the cached signature. If not cached, the signature is just appended to a StringBuilder but we need to convert that to a String when we want to cache it.

A few ideas:
- When we hit a limit of total cache size, just disable the cache, assuming the application creates a lot of dynamic SQLs.
- Don't use a weak map and just cache the first x SQLs. Hit rate may not be great though.

Reply:
@felixbarny do you mean just stop caching new statements once you've hit 200? Seems fine, but then you've still got to deal with concurrent map updates.

I'm not really convinced that scanning each time is going to be a problem in practice. Applications that make repeat queries should be using prepared statements. If they're not, then I would take that as either an indication of one of the following:

 - performance of that particular bit of code is either cold / unimportant
 - it's hot and the devs don't know it; it will be (slightly?) exacerbated, but also highlighted by the agent as something to fix
 - it's a dynamically generated query, so will end up overflowing the cache anyway, and cache miss for every >200th query

So... I'd just benchmark it and see whether it's worth it.

I just went and benchmarked the base overhead for Go:

```
BenchmarkStmtQueryContext/baseline-8                      200000              9332 ns/op            1008 B/op         16 allocs/op
BenchmarkStmtQueryContext/baseline_prepared-8             300000              4790 ns/op             912 B/op         10 allocs/op
BenchmarkStmtQueryContext/instrumented-8                  200000              9213 ns/op             822 B/op         16 allocs/op
BenchmarkStmtQueryContext/instrumented_prepared-8         300000              4885 ns/op             721 B/op         10 allocs/op
BenchmarkQuerySignature-8                                1000000              1552 ns/op          77.94 MB/s          32 B/op          1 allocs/op
```

The first four are repeatedly executing "SELECT * FROM foo" against an in-memory SQLite database. The "prepared" variants use a prepared statement, the others do not. In either case, the query is short and simple, and the overhead is ~100ns/query.

The last benchmark is just repeatedly scanning a longer query: "SELECT *,(SELECT COUNT(*) FROM table2 WHERE table2.field1 = table1.id) AS count FROM table1 WHERE table1.field1 = 'value'", and we can see the overhead there is more substantial.

You would have to benchmark against using a concurrent hash map to know if it's worth it, and at what query length threshold.

Reply:
> do you mean just stop caching new statements once you've hit 200
That's option 2 which only really makes sense when not using weak keys.

Option 1 is to really deactivate caching altogether. The reasoning is that when using a weak reference to the statement we may do a lot of allocations, as described in my previous comment.

Reply:
> Applications that make repeat queries should be using prepared statements. 

The current code will still run every time. Do you suggest caching for the PreparedStatement objects instead? This is an option and we actually have such cache already. However, it will cover less cases.

I assume most applications use repeated queries, so I do believe this is something we should do in one for or another.

> When we hit a limit of total cache size, just disable the cache, assuming the application creates a lot of dynamic SQLs.

This can work, just make sure you either count yourself, or you count removals, as the weak-key maps clean GC keys, so map size is not a good indication.

Reply:
> That's option 2 which only really makes sense when not using weak keys.
>
> Option 1 is to really deactivate caching altogether. The reasoning is that when using a weak reference to the statement we may do a lot of allocations, as described in my previous comment.

@felixbarny sorry, I was replying to your previous message - hadn't refreshed and seen your follow up. Makes sense.

> The current code will still run every time. Do you suggest caching for the PreparedStatement objects instead? This is an option and we actually have such cache already. However, it will cover less cases.

@eyalkoren ah, I had meant to check if prepared statements were being cached, but forgot. This is essentially what I do in Go, and it works well. I have to override the prepared statement object anyway, so it's natural to attach the span name to the overridden prepared statement object. I don't know what the performance implication for the Java agent would be, so yet another thing to compare.

Reply:
> This can work, just make sure you either count yourself,

What do you mean with "count yourself"? I'm only one 😜 

> or you count removals, as the weak-key maps clean GC keys, so map size is not a good indication.

Do you mean that after x keys have been GC'd, we should stop caching?

Only caching the span name for prepared statements sounds good. But prepared statements queries may still be constructed dynamically, causing unnecessary allocations for the span name. For example when there's a search which multiple inputs which correspond to different properties (first name, last name, age range, ...) the query may be constructed based on which fields are filled out but still uses prepared statements to leverage parameter placeholders.